New `docker daemon` command

tiborvass opened the pull request:
Fixes #9677

This patch creates a new cli package that allows to combine both client
and daemon commands (there is only one daemon command: docker daemon).

The `-d` and `--daemon` top-level flags are deprecated and a special
message is added to prompt the user to use `docker daemon`.

Providing top-level daemon-specific flags for client commands result
in an error message prompting the user to use `docker daemon`.

This patch does not break any old but correct usages.

Signed-off-by: Tibor Vass tibor@docker.com

Thanks to @rhatdan and @shishir-a412ed for their huge patience and their initial work.

Ref #6055 #9255 #10535 #11418

Commit:
Add and modify tests for legacy and new daemon invokations

Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>
Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
cli: new daemon command and new cli package

This patch creates a new cli package that allows to combine both client
and daemon commands (there is only one daemon command: docker daemon).

The `-d` and `--daemon` top-level flags are deprecated and a special
message is added to prompt the user to use `docker daemon`.

Providing top-level daemon-specific flags for client commands result
in an error message prompting the user to use `docker daemon`.

This patch does not break any old but correct usages.

This also makes `-d` and `--daemon` flags, as well as the `daemon`
command illegal in client-only binaries.

Signed-off-by: Tibor Vass <tibor@docker.com>

icecrime commented:
Design was approved in the preceding PR, and it seems to be in line with the earlier discussions, so I suggest we directly go to code review.

thaJeztah commented:
ping @albers for reviewing completion scripts :-) :heart:

albers commented on contrib/completion/bash/docker:
```diff
@@ -201,19 +201,36 @@ __docker_signals() {
+		--tlsverify
+		--tlscacert
+		--tlscert
+		--tlskey
```
The preceding three options are not boolean and should call `_filedir`

tiborvass replied:
I don't know much about `_filedir` would you mind showing me a fix?

albers replied:
See the completion for `--pidfile|-p|--tlscacert|--tlscert|--tlskey)` in `_docker_daemon()`.
By the way. if you would prefer to concentrate on real coding stuff, I could take over the bash completion part. Just give me a ping.

albers replied:
`--tlscacert|--tlscert|--tlskey` should also be removed from `_docker_daemon()`

albers commented on contrib/completion/bash/docker:
```diff
@@ -201,19 +201,36 @@ __docker_signals() {
+
+	case "$cur" in
+		-*)
+			COMPREPLY=( $( compgen -W "$boolean_options $main_options_with_args" -- "$cur" ) )
```
`$main_options_with_args` still contains the previous set of top-level arguments, including those needed for daemon mode. Therefore, completion will wrongly support things like `docker --api-cors-header`.
Please reduce the definition of `$main_options_with_args` below to the actual global options that take args: 

```
--host -H --log-level -l --tlscacert --tlscert --tlskey
```

albers commented on contrib/completion/bash/docker:
```diff
@@ -201,19 +201,36 @@ __docker_signals() {
 		--selinux-enabled
-		--tls
-		--tlsverify
 		--userland-proxy=false
```
I suggest to also modify completion for the others boolean options defaulting to `true` so that they also complete to `--xxx=false`:
```
--icc --ip-forward --ip-masq --iptables
```

albers commented on contrib/completion/bash/docker:
```diff
@@ -246,9 +263,6 @@ _docker_docker() {
 		-*)
 			COMPREPLY=( $( compgen -W "$boolean_options $main_options_with_args" -- "$cur" ) )
```
`$main_options_with_args` does not belong here as it contains (part of the) global flags like `-D`.
Please remove the non-global flags from `$main_options_with_args` below (see comment above) and insert them here literally or via a local variable.

albers commented:
@thaJeztah I added some comments and I offer to take over the bash completion part. It's a bit more complex than usual and quite hacky.
I cannot comment on the zsh part because I'm not yet familiar with it.

thaJeztah commented:
Thanks so much @albers, I knew this was quite a big change, so happy to have you on board here!

@tiborvass what's the easiest way for @albers to carry the bash completion?

tiborvass commented:
@thaJeztah I can remove it or leave it as-is for now until we have some more review on the docker side of things.

albers commented:
I think it's better to remove the completion part here. It will leave the remaining part more focussed and my follow up PR more self-contained.
I will also address your addition of the zfs storage driver in another PR.

thaJeztah commented:
Thanks @albers!

calavera commented on api/client/help.go:
```diff
@@ -1,5 +1,6 @@
 package client
 
+/*
```
dead code?

calavera replied:
@tiborvass can we remove this file?

calavera commented:
```
root@d# ./docker daemon
INFO[0000] [graphdriver] using prior storage driver "aufs"
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock)
WARN[0000] Running modprobe bridge nf_nat failed with message: , error: exit status 1
WARN[0000] Your kernel does not support swap memory limit.
INFO[0000] Loading containers: start.
............
INFO[0000] Loading containers: done.
INFO[0000] Daemon has completed initialization
```

:heart_eyes_cat: LGTM after a rebase :wink:

thaJeztah commented on contrib/vagrant-docker/README.md:
```diff
@@ -31,7 +31,7 @@ stop on runlevel [!2345]
 
 script
-    /usr/bin/docker -d -H=tcp://0.0.0.0:2375
+    /usr/bin/docker -H=tcp://0.0.0.0:2375 daemon
```
does `docker daemon -H=tcp://0.0.0.0:2375` also work?

albers replied:
Aside from the technical aspect, IMHO the `=arg`syntax is confusing when used on a short option. It is very common for the long form options but I cannot recall having seen this elsewhere before.
I recommend using either `-H tcp://0.0.0.0:2375` or `--host=tcp://0.0.0.0:2375`

duglin commented:
rebase needed

duglin commented:
and probably a squash

tiborvass commented on cli/cli.go:
```diff
@@ -0,0 +1,203 @@
+		cli.noSuchCommand(args[0])
+	}
+
+	if cli == nil {
```
should be `cli.Usage == nil`, will fix.

tiborvass commented on docker/daemon_none.go:
```diff
@@ -0,0 +1,19 @@
+// +build !daemon
```
basically this file's older version was named `docker/client.go`: https://github.com/docker/docker/pull/13771/files#diff-210c31cd49f34ac9a1fcc99dd77d9030L1

thaJeztah commented:
@tiborvass is this still on track for 1.8? Do we need more reviewers here?

calavera commented:
@thaJeztah yes, we're still aiming to release this in 1.8.

tiborvass commented:
@thaJeztah yes

tiborvass commented:
test failures are unrelated

calavera commented on cli/cli.go:
```diff
@@ -0,0 +1,203 @@
+	// make the generic Cli object the first cli handler
+	// in order to handle `docker help` appropriately
+	cli := new(Cli)
+	handlers = append(handlers, nil)
```
these 4 lines could be `cli.handlers = append([]Handler{cli}, handlers...)`

calavera commented on cli/cli.go:
```diff
@@ -0,0 +1,203 @@
+			return err.error
+		}
+	}
+	if len(args) > 0 {
```
is this `if len(args) == 1`?

tiborvass replied:
no, because it's also the codepath for `command not found` when `err` is not an `initErr`.

calavera replied:
oh gotcha, thank you.

calavera commented:
That `Cli` named import looks kinda weird, but I can live with it, I think.

tiborvass commented:
@calavera yeah it's super weird, but it was there for simpler codereview. We could always have a followup cleanup PR, I just didn't want to make this longer to review.

calavera commented:
> We could always have a followup cleanup PR, I just didn't want to make this longer to review.

I'm okay with that :+1:

tiborvass commented:
Not sure why janky failed on TestSearchCmdOptions... race condition? It works fine locally and it's the first time I see this error on this PR.

tiborvass commented:
rebased

estesp commented:
The Windows test failure looks like something that should actually be "ok"--meaning a client-only binary **should** output this message when trying to run a daemon.  So I think the test needs to be skipped or at least handle the case if binary == client-only:
```
FAIL: docker_cli_help_test.go:14: DockerSuite.TestHelpTextVerify

docker_cli_help_test.go:116:
    c.Fatalf("Error on %q help. non-empty stderr:%q", cmd, stderr)
... Error: Error on "daemon" help. non-empty stderr:"2015/07/22 00:20:54 This is a client-only binary - running the Docker daemon is not supported.\n"
```

tiborvass commented:
@estesp yes that one I think I have a fix but the other one is trickier (execID). If you know what the issue is I'm all ears :)

calavera commented on pkg/mflag/flag.go:
```diff
@@ -1143,3 +1143,56 @@ func (f *FlagSet) Init(name string, errorHandling ErrorHandling) {
+					msg = fmt.Sprintf("%s flag redefined: %s", fset.name, k)
+				}
+				fmt.Fprintln(fset.Out(), msg)
+				panic(msg) // Happens only if flags are declared with identical names
```
can we not panic? what if we check the names before changing the destination and return an error when there are duplicated names?

tiborvass replied:
i found it strange too, but I copied it from another part of the lib that was already doing this. It's also in the stdlib

tiborvass replied:
https://golang.org/src/flag/flag.go#L679

calavera replied:
fortunately for us, we don't need to keep the behavior backwards compatible, do we? :wink:

tiborvass commented on pkg/mflag/flag.go:
```diff
@@ -1143,3 +1143,56 @@ func (f *FlagSet) Init(name string, errorHandling ErrorHandling) {
 	f.errorHandling = errorHandling
 }
+
+func (f *FlagSet) DeprecateAll() {
```
this shouldn't be here... weird

thaJeztah commented:
just so I don't forget later;

ping @jfrazelle to keep an eye on this, as I think the [init scripts](https://github.com/docker/docker/tree/master/contrib/init) will need to be updated after this is merged

calavera commented:
@thaJeztah I've updated the init scripts in #14865

thaJeztah commented:
Thanks @calavera !

estesp commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -324,6 +324,90 @@ func (s *DockerDaemonSuite) TestDaemonLogLevelWrong(c *check.C) {
 	c.Assert(s.d.Start("--log-level=bogus"), check.NotNil, check.Commentf("Daemon shouldn't start with wrong log level"))
 }
 
+func (s *DockerSuite) TestDaemonStartWithBackwardCompatibility(c *check.C) {
```
These should be in the `*DockerDaemonSuite`, not `*DockerSuite`, right?

tiborvass replied:
No because we need to manually handle the daemon object. We run NewDaemon in the loop, and then need to set GlobalFlags for the invalid part

estesp replied:
Ah right.. sorry, ignore me :)

cpuguy83 commented on integration-cli/docker_cli_exec_test.go:
```diff
@@ -339,6 +339,7 @@ func (s *DockerSuite) TestInspectExecID(c *check.C) {
 
 		out = strings.TrimSuffix(out, "\n")
 		if out != "[]" && out != "<no value>" {
+			c.Logf("out: %q", out)
```
Does this log need to be here? Looks like you possibly had this here for test debugging?

tiborvass replied:
not it was when i tried to debug windows ci issues :)

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -217,6 +217,18 @@ func (s *DockerSuite) TestRunLinksContainerWithContainerId(c *check.C) {
 
+// Issue 9677.
+func (s *DockerSuite) TestRunWithDaemonFlags(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "--selinux-enabled", "run", "-i", "-t", "busybox", "true")
```
`dockerCmdWithError`

Commit:
Update docs to use the new `docker daemon` command

Signed-off-by: Tibor Vass <tibor@docker.com>

cpuguy83 commented:
LGTM

duglin commented:
Is there anything we can do about having api/client/* ; cli/* and docker/*    ?
There seems to be some confusion w.r.t. where the real docker cli is.
I always thought that docker/* was the main entrypoint and then it would go into either docker/client.go or docker/daemon.go.  But I'm not sure how the new "cli" dir fits into it.  If its common stuff then it seems like it should go into docker/, but if its client specific then api/client seems like the right spot.
Or an I missing something?

LK4D4 commented:
I got panic:
```
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/docker/docker/opts.(*MapOpts).Set(0xc208060d80, 0x7ffd3a82d309, 0xb, 0x0, 0x0)
        /go/src/github.com/docker/docker/opts/opts.go:185 +0x1ce
github.com/docker/docker/pkg/mflag.(*FlagSet).parseOne(0xc208058500, 0xc208000101, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/mflag/flag.go:994 +0x14e5
github.com/docker/docker/pkg/mflag.(*FlagSet).Parse(0xc208058500, 0xc208000130, 0x10, 0x10, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/mflag/flag.go:1029 +0xa1
github.com/docker/docker/pkg/mflag.Parse()
        /go/src/github.com/docker/docker/pkg/mflag/flag.go:1116 +0x79
main.main()
        /go/src/github.com/docker/docker/docker/docker.go:51 +0x27f
```
with options:
```
DEBUG=1 docker -d -D -s overlay -H unix:///var/run/docker.sock -H
tcp://127.0.0.1:5001 --ipv6 --fixed-cidr-v6=2001::/56 --userland-proxy=false --log-driver=json-file --log-opt max-size=1k --exec-opt native.cgroupdriver=cgroupfs
```

calavera commented:
@duglin I think `cli/*` is fine because it's a common package for the other two, but I'm open to suggestions.

I just tested it and the error that @LK4D4 is gone.

LGTM

tiborvass commented:
@LK4D4 fixed panic

@duglin the problem is that before this PR, the global flags code was in `docker/` and the cli handling was in `api/client/`. I took the choice of changing as few lines of code as possible to make this work. We can always refactor more later. The `cli/` package is there because it's combining two cli handlers (one for client, one for daemon). An unfortunate thing it does too is hold the state for CommonFlags and ClientFlags, the reason is it's because I had to send it to `api/client` because that's where currently the client cli is initialized. I do welcome all kinds of refactoring.

LK4D4 commented:
It works for me.

duglin commented:
@calavera I was thinking that `cli/*` could be moved into `docker/` since both seem to contains files that are shared/common to both the cli and daemon.

calavera commented:
I don't have a strong opinion, either is fine to me.

tiborvass commented:
@duglin i wanted that, but i couldn't because `api/client` needs it.

duglin commented:
@tiborvass should we official deprecate "-d" by mentioning it on the "deprecated features" page? #13900 was just merged

duglin commented on docker/daemon.go:
```diff
@@ -26,17 +29,26 @@ import (
+
+// NewDaemonCli returns a pre-configured daemon CLI
+func NewDaemonCli() *DaemonCli {
+	daemonFlags = cli.Subcmd("daemon", nil, "Enable daemon mode", true)
```
I know this is from the old code but `Enable daemon mode` sounds odd. Almost like its turning on daemon mode in addition to its real/main task.  What about `Start the Docker daemon`  ?  This is repeated down below too.

duglin replied:
Did you want to address this one?

duglin replied:
Note: we think this text isn't shown any more

tiborvass commented:
I'm fine with any docs/help suggestions

duglin commented on integration-cli/docker_cli_help_test.go:
```diff
@@ -103,6 +103,9 @@ func (s *DockerSuite) TestHelpTextVerify(c *check.C) {
 			// Grab just the first word of each line
 			cmd = strings.Split(strings.TrimSpace(cmd), " ")[0]
 			cmds = append(cmds, cmd)
+			if cmd == "daemon" {
```
I don't think this should be here - it should be added to `noShortUsage` instead.

duglin replied:
In particular there are lots of tests we need to run on it - like --help, --bad-arg...

duglin replied:
I see that you added it below - I just removed the "continue" and the test passes so you should be good to remove these lines.

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -217,6 +217,17 @@ func (s *DockerSuite) TestRunLinksContainerWithContainerId(c *check.C) {
+	out, _, err := dockerCmdWithError(c, "--selinux-enabled", "run", "-i", "-t", "busybox", "true")
+	if err != nil {
+		if !strings.Contains(out, "must follow the 'docker daemon' command") && // daemon
+			!strings.Contains(out, "flag provided but not defined: --selinux-enabled") { // no daemon (client-only)
```
Just curious - do we ever run the tests with just the cli-only version of the docker exe?

tianon replied:
Yes, although currently only on Windows.

calavera commented:
@duglin, @tiborvass I've added the flag to the deprecated features in https://github.com/docker/docker/pull/14877

duglin commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -324,6 +324,90 @@ func (s *DockerDaemonSuite) TestDaemonLogLevelWrong(c *check.C) {
+		daemon
+	)
+
+	var flags = []map[kind][]string{
```
Why do we support certain flags to be global/common as well as daemon specific?

tiborvass replied:
Because the `common` ones are for the client.

duglin commented on docker/daemon.go:
```diff
@@ -79,14 +91,37 @@ func migrateKey() (err error) {
+		logrus.Warn("please use 'docker daemon' instead.")
+	} else if !(commonFlags.FlagSet).IsEmpty() {
+		// deny `docker -D daemon`
+		logrus.Fatal("invalid global flags")
```
This error isn't helpful - can we say what's wrong?
I'm seeing this:
```
$ docker --log-level=debug daemon
FATA[0000] invalid global flags
```
but according to the help text it should be legal.

tiborvass replied:
this should not be legal

duglin replied:
Ignoring whether --log-level should be global or not, the error message isn't helpful because it doesn't tell the user what flag was bad.  If I had listed several of them I wouldn't know which one was wrong - especially since --log-level (per the help) is legal.  Can't we show the erroneous flag in the msg?

tiborvass replied:
@duglin we could do a `Visit()` on `commonFlags` which would visit all "global" flags (except -v and -h) that were set by the user. Do you think we should do that? Show all flags or only the first one or only the last one?

duglin replied:
if calling Visit() will cause a nicer error message then probably. Don't understand the code well enough yet to know the right code solution.

duglin commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -324,6 +324,90 @@ func (s *DockerDaemonSuite) TestDaemonLogLevelWrong(c *check.C) {
+		d.Stop()
+	}
+
+	// `docker -l info daemon --selinux-enabled`
```
this is really confusing to see `-l` listed as a valid global options but then to have `docker daemon` flag it as an error.  Why do we do this?

tiborvass replied:
naming is the hardest, they used to be global flags, i refer them as "common" flags, because they are common to client and daemon, but they are not global flags anymore (only for backwards compat with `-d` but then they are not in any help text).

The main reason I had to change the design is because as soon as you discriminate between global flags and daemon flags, you force all the init scripts to expose a $DAEMON_OPTS envvar in addition to an already existing $DOCKER_OPTS envvar, otherwise you can write neither `docker daemon $DOCKER_OPTS` nor `docker $DOCKER_OPTS daemon`.

So we decided to have `docker daemon $DOCKER_OPTS` and the only truly global flags are `-h` and `-v` which are handled very early on and exit.

duglin replied:
This is really confusing.  Looking at it from a user's point of view, they'll see:
```
Usage: docker [OPTIONS] COMMAND [arg...]

A self-sufficient runtime for containers.

Options:

  --config=~/.docker              Location of client config files
  -D, --debug=false               Enable debug mode
  -H, --host=[]                   Daemon socket(s) to connect to
  -h, --help=false                Print usage
  -l, --log-level=info            Set the logging level
...
Commands:
    attach    Attach to a running container
    build     Build an image from a Dockerfile
    commit    Create a new image from a container's changes
    cp        Copy files/folders from a container to a HOSTDIR or to STDOUT
    create    Create a new container
    daemon    Enable daemon mode
...
```
which should mean:
```
docker --log-level=debug daemon
```
should be legal.  We need to either fix the code to match the help, or fix the help to match the code.

tiborvass replied:
@duglin that's for the "client-side". Do you think we should have `[CLIENT-OPTIONS]` ?

duglin replied:
that might be the easiest fix - OPTIONS -> CLIENT_OPTIONS  and Options -> Client-Options.  Would need to see how it looks.

I wonder if we need to add something to the 'daemon' line to make it clear its not a client command. Should be obvious but ya never know....   (guess it depends on whether it can fit in 80 chars).  :)

tiborvass replied:
@duglin I suck at expressing myself correctly, so feel free to tell me what to put and I'll do it :D

The only problem with `[CLIENT-OPTIONS]` is that `-h` and `-v` are really global. So maybe we should keep `[OPTIONS]` and have `Global Options:` as well as `Client Options:` ?

We could add `See docker daemon --help` in the line where there's the `daemon` command, wdyt?

duglin commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -1211,15 +1294,14 @@ func (s *DockerDaemonSuite) TestDaemonRestartKillWait(c *check.C) {
 func (s *DockerDaemonSuite) TestHttpsInfo(c *check.C) {
 	const (
 		testDaemonHTTPSAddr = "localhost:4271"
+		host                = "tcp://" + testDaemonHTTPSAddr
```
not sure I see the point in having 2 vars when the 1st is only used in the 2nd.  Why not just add "tcp://" to testDaemonHTTPSAddr ?

tiborvass replied:
you're right

tiborvass commented:
I added the following changes:
- Modified the synopsis (removed `daemon` in the list of commands and instead added multiple lines for Usage)
- Breaking change/fix: `docker -d` or `docker daemon` are invalid in client-only binaries (also, they are not present in the synopsis for client-only binaries)
- Made `docker --config /tmp daemon` fail and updated a test for it.

calavera commented:
It looks like global flags work now as expected:

```
~$ sudo ./docker-1.8.0-dev -H 0.0.0.0:3456 daemon
FATA[0000] invalid global flags
~$ sudo ./docker-1.8.0-dev daemon -H 0.0.0.0:3456
WARN[0000] Udev sync is not supported. This will lead to unexpected behavior, data loss and errors. For more information, see https://docs.docker.com/reference/commandline/cli/#daemon-storage-driver-option
WARN[0000] /!\ DON'T BIND ON ANY IP ADDRESS WITHOUT setting -tlsverify IF YOU DON'T KNOW WHAT YOU'RE DOING /!\
INFO[0000] Listening for HTTP on tcp (0.0.0.0:3456)
```

The error on windows is not related to this PR. It's solved in #14925.

calavera commented:
@tiborvass can we get those commits squashed or with better descriptions?

tiborvass commented:
@calavera sure

calavera commented:
:heart_eyes_cat: 

```
~$ sudo ./docker-1.8.0-dev daemon -H 0.0.0.0:3456
2015/07/24 00:03:45 This is a client-only binary - running the Docker daemon is not supported.
```

tiborvass commented:
@calavera that shows you're not testing this PR :)
```
$ docker daemon -H 0.0.0.0:3456
docker: 'daemon' is not a docker command.
See 'docker --help'.
```

calavera commented:
oh, much better! thanks

calavera commented:
The new changes also LGTM

duglin commented on cli/cli.go:
```diff
@@ -0,0 +1,200 @@
+		switch err := err.(type) {
+		case nil:
+			return command(args[2:]...)
+		case initErr:
```
what if its not an initErr ?

tiborvass replied:
then it moves on to `if len(args) > 0`

duglin replied:
well yea, I got that.  Just expected it to do something with it. I guess we'll deal with it at line 93.

tiborvass replied:
I kept the logic from before: https://github.com/docker/docker/blob/37209190c76de66531acba848f1537da2899e32d/api/client/cli.go#L90

duglin commented:
way too much code to review :-)   but it seems good. We can fix any bugs that pop-up later.
LGTM!

jfrazelle commented:
no going back now ;)

shishir-a412ed commented:
@tiborvass 
:clap: :clap: :clap: :clap:

thaJeztah commented:
Oh yes!!

I did a quick scan of the docs changes and the *changes* LGTM, I hope we got them all :-)

thaJeztah commented:
ping @albers - hope you are able to work on the completion scripts now :smile:

albers commented:
It will take about one week until I can get back to it. I'm reopening #13934 to indicate that it's in progress.

thaJeztah commented:
Thanks for the heads-up, @albers!

rhatdan commented:
Finally, this might have been one of the Oldest Patchsets/requests.  Thanks.

thaJeztah commented:
Thank you @rhatdan, IIRC, you started this PR-relay originally :+1:

albers commented:
@sdurrheimer Hi Steve, docker -d was replaced by the new `docker daemon` command. Maybe you could add this to zsh completion?

thaJeztah commented:
Think we need the fish completions updated as well; I think those are auto-generated
